File: design_tracker.py
#!/usr/bin/env python3
"""DesignArena registry tracker — diffs /api/registry and alerts on Telegram."""

import json
import os
import random
import time
from datetime import datetime, timezone

import requests

URL = "https://www.designarena.ai/api/registry"
SNAPSHOT = "design_snapshot.json"

TG_TOKEN = os.environ.get("TELEGRAM_TOKEN", "").strip()
TG_CHAT = os.environ.get("TELEGRAM_CHAT_ID", "").strip()

UA = ("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
      "(KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36")


# ---------- fetch ----------

def fetch_registry():
    last = "unknown"
    for attempt in range(6):
        try:
            r = requests.get(URL, headers={"User-Agent": UA, "Accept": "application/json"}, timeout=60)
            if r.status_code == 200:
                data = r.json()
                if isinstance(data, dict) and "models" in data:
                    print(f"fetch OK on attempt {attempt + 1}")
                    return data
                last = "no 'models' key in JSON"
            else:
                last = f"HTTP {r.status_code}"
        except Exception as e:
            last = str(e)
        wait = 5 + random.random() * 10
        print(f"attempt {attempt + 1} failed ({last}), retrying in {wait:.0f}s")
        time.sleep(wait)
    raise RuntimeError(f"fetch failed: {last}")


# ---------- normalize ----------

def norm_models(raw):
    out = {}
    for mid, m in raw.items():
        arenas = m.get("arenas") or {}
        out[mid] = {
            "displayName": m.get("displayName") or mid,
            "provider": m.get("provider"),
            "active": bool(m.get("active")),
            "openSource": bool(m.get("openSource")),
            "router": bool(m.get("router")),
            "vision": bool(m.get("vision")),
            "inputModalities": sorted(m.get("inputModalities") or []),
            "arenas": {k: sorted(v) for k, v in arenas.items() if isinstance(v, list)},
        }
    return out


def norm_providers(raw):
    return {pid: (p or {}).get("displayName") or pid for pid, p in (raw or {}).items()}


def norm_pricing(raw):
    """flatten pricing to {model_id: {'text.input': 1.5, 'image.perImage': 0.04, ...}}"""
    out = {}
    for mid, cats in (raw or {}).items():
        flat = {}
        for cat, fields in (cats or {}).items():
            for k, v in (fields or {}).items():
                flat[f"{cat}.{k}"] = v
        if flat:
            out[mid] = flat
    return out


# ---------- snapshot ----------

def load_snapshot():
    if not os.path.exists(SNAPSHOT):
        return None
    with open(SNAPSHOT, encoding="utf-8") as f:
        return json.load(f)


def save_snapshot(models, providers, pricing):
    with open(SNAPSHOT, "w", encoding="utf-8") as f:
        json.dump({"models": models, "providers": providers, "pricing": pricing},
                  f, indent=2, sort_keys=True)


# ---------- diff ----------

def diff(old, new):
    om, op, opz = old["models"], old.get("providers", {}), old.get("pricing", {})
    nm, np_, npz = new["models"], new.get("providers", {}), new.get("pricing", {})

    d = {"added": [], "removed": [], "renamed": [], "provider": [], "active_on": [],
         "active_off": [], "categories": [], "caps": [], "new_providers": [],
         "price_new": [], "price_changed": []}

    for mid in sorted(set(nm) - set(om)):
        m = nm[mid]
        cats = ", ".join(f"{k}: {', '.join(v)}" for k, v in m["arenas"].items()) or "none"
        d["added"].append((mid, m, cats))

    for mid in sorted(set(om) - set(nm)):
        d["removed"].append((mid, om[mid]))

    for mid in sorted(set(om) & set(nm)):
        o, n = om[mid], nm[mid]
        if o["displayName"] != n["displayName"]:
            d["renamed"].append((mid, o["displayName"], n["displayName"]))
        if o["provider"] != n["provider"]:
            d["provider"].append((mid, n["displayName"], o["provider"], n["provider"]))
        if o["active"] != n["active"]:
            (d["active_on"] if n["active"] else d["active_off"]).append((mid, n["displayName"]))

        # category (arena) changes
        cat_changes = []
        for cat in sorted(set(o["arenas"]) | set(n["arenas"])):
            added = sorted(set(n["arenas"].get(cat, [])) - set(o["arenas"].get(cat, [])))
            removed = sorted(set(o["arenas"].get(cat, [])) - set(n["arenas"].get(cat, [])))
            if added:
                cat_changes.append(f"➕ {cat}: {', '.join(added)}")
            if removed:
                cat_changes.append(f"➖ {cat}: {', '.join(removed)}")
        if cat_changes:
            d["categories"].append((mid, n["displayName"], cat_changes))

        # capability changes
        cap_changes = []
        for field, label in [("openSource", "openSource"), ("router", "router"), ("vision", "vision")]:
            if o[field] != n[field]:
                cap_changes.append(f"{label}: {o[field]} ➡️ {n[field]}")
        if o["inputModalities"] != n["inputModalities"]:
            cap_changes.append(f"inputModalities: {', '.join(o['inputModalities']) or 'none'} "
                               f"➡️ {', '.join(n['inputModalities']) or 'none'}")
        if cap_changes:
            d["caps"].append((mid, n["displayName"], cap_changes))

    for pid in sorted(set(np_) - set(op)):
        d["new_providers"].append((pid, np_[pid]))

    for mid in sorted(set(npz) - set(opz)):
        d["price_new"].append((mid, npz[mid]))
    for mid in sorted(set(npz) & set(opz)):
        changes = []
        for k in sorted(set(npz[mid]) | set(opz[mid])):
            ov, nv = opz[mid].get(k), npz[mid].get(k)
            if ov != nv:
                changes.append(f"{k}: {ov} ➡️ {nv}")
        if changes:
            d["price_changed"].append((mid, changes))

    return d


def has_changes(d):
    return any(d.values())


# ---------- message ----------

LIMIT = 15

def build_message(d, models):
    now = datetime.now(timezone.utc).strftime("%d %b %Y, %H:%M UTC")
    lines = [f"🎨 DesignArena Tracker — {now}", ""]

    def section(title, items, render):
        if not items:
            return
        lines.append(f"{title} ({len(items)})")
        for it in items[:LIMIT]:
            lines.extend(render(it))
        if len(items) > LIMIT:
            lines.append(f"… +{len(items) - LIMIT} more")
        lines.append("")

    section("🆕 New models", d["added"], lambda it: [
        f"• {it[1]['displayName']} ({it[0]})",
        f"  provider: {it[1]['provider']} | active: {'✅' if it[1]['active'] else '❌'} "
        f"| open source: {'✅' if it[1]['openSource'] else '❌'}",
        f"  categories: {it[2]}",
    ])
    section("❌ Removed models", d["removed"], lambda it: [
        f"• {it[1]['displayName']} ({it[0]}) — provider: {it[1]['provider']}",
    ])
    section("✏️ Name updates", d["renamed"], lambda it: [
        f"• {it[1]} ➡️ {it[2]}", f"  {it[0]}",
    ])
    section("🏢 Provider updates", d["provider"], lambda it: [
        f"• {it[1]} ({it[0]}): {it[2]} ➡️ {it[3]}",
    ])
    section("🟢 Activated (now live)", d["active_on"], lambda it: [
        f"• {it[1]} ({it[0]})",
    ])
    section("🔴 Deactivated (pulled)", d["active_off"], lambda it: [
        f"• {it[1]} ({it[0]})",
    ])
    section("🏟️ Category updates", d["categories"], lambda it: [
        f"• {it[1]} ({it[0]})", *[f"  {c}" for c in it[2]],
    ])
    section("⚡ Capability updates", d["caps"], lambda it: [
        f"• {it[1]} ({it[0]})", *[f"  {c}" for c in it[2]],
    ])
    section("🏭 New providers", d["new_providers"], lambda it: [
        f"• {it[1]} ({it[0]})",
    ])
    section("💲 Pricing added", d["price_new"], lambda it: [
        f"• {it[0]}: " + ", ".join(f"{k}={v}" for k, v in list(it[1].items())[:4]),
    ])
    section("💲 Pricing updates", d["price_changed"], lambda it: [
        f"• {it[0]}", *[f"  {c}" for c in it[1]],
    ])

    return "\n".join(lines).strip()


# ---------- telegram ----------

def send_telegram(text):
    if not TG_TOKEN or not TG_CHAT:
        print("--- dry run (no Telegram secrets) ---")
        print(text)
        return
    for i in range(0, len(text), 4000):
        chunk = text[i:i + 4000]
        r = requests.post(
            f"https://api.telegram.org/bot{TG_TOKEN}/sendMessage",
            json={"chat_id": TG_CHAT, "text": chunk, "disable_web_page_preview": True},
            timeout=30,
        )
        if r.status_code != 200:
            print("telegram error:", r.status_code, r.text[:300])
        time.sleep(1)
    print("telegram sent")


# ---------- main ----------

def main():
    old = load_snapshot()
    data = fetch_registry()
    new = {
        "models": norm_models(data.get("models", {})),
        "providers": norm_providers(data.get("providers", {})),
        "pricing": norm_pricing(data.get("pricing", {})),
    }
    print(f"models: {len(new['models'])}, providers: {len(new['providers'])}, "
          f"priced: {len(new['pricing'])}")

    if old is None:
        save_snapshot(**new)
        print("baseline snapshot saved — no alert (first run)")
        return

    d = diff(old, new)
    if not has_changes(d):
        print("no changes")
        return

    msg = build_message(d, new["models"])
    send_telegram(msg)
    save_snapshot(**new)
    print("snapshot updated")


if __name__ == "__main__":
    main()

File: design_tracker.mjs
#!/usr/bin/env node
/**
 * DesignArena Registry Tracker — Native Node.js Edition
 * Diffs https://www.designarena.ai/api/registry and alerts Discord instantly.
 */

import fs from 'node:fs';

const URL = 'https://www.designarena.ai/api/registry';
const SNAPSHOT = 'design_snapshot.json';

const DISCORD_WEBHOOK_URL = (
  process.env.DISCORD_WEBHOOK_URL ||
  process.env.DISCORD_WEBHOOK ||
  ''
).trim();

async function fetchRegistry() {
  const headers = {
    'User-Agent':
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36',
    Accept: 'application/json',
  };

  for (let attempt = 1; attempt <= 4; attempt++) {
    try {
      const res = await fetch(URL, { headers, signal: AbortSignal.timeout(15000) });
      if (res.ok) {
        const data = await res.json();
        if (data && typeof data === 'object' && data.models) {
          return data;
        }
      }
    } catch (err) {
      console.warn(`Attempt ${attempt} failed: ${err.message}`);
    }
    await new Promise((r) => setTimeout(r, 1000));
  }
  throw new Error('Failed to fetch DesignArena registry');
}

function normModels(raw = {}) {
  const out = {};
  for (const [mid, m] of Object.entries(raw)) {
    const arenas = m?.arenas || {};
    const normArenas = {};
    for (const [k, v] of Object.entries(arenas)) {
      if (Array.isArray(v)) normArenas[k] = [...v].sort();
    }
    out[mid] = {
      displayName: m?.displayName || mid,
      provider: m?.provider || null,
      active: Boolean(m?.active),
      openSource: Boolean(m?.openSource),
      router: Boolean(m?.router),
      vision: Boolean(m?.vision),
      inputModalities: [...(m?.inputModalities || [])].sort(),
      arenas: normArenas,
    };
  }
  return out;
}

function normProviders(raw = {}) {
  const out = {};
  for (const [pid, p] of Object.entries(raw)) {
    out[pid] = p?.displayName || pid;
  }
  return out;
}

function normPricing(raw = {}) {
  const out = {};
  for (const [mid, cats] of Object.entries(raw)) {
    const flat = {};
    for (const [cat, fields] of Object.entries(cats || {})) {
      for (const [k, v] of Object.entries(fields || {})) {
        flat[`${cat}.${k}`] = v;
      }
    }
    if (Object.keys(flat).length > 0) out[mid] = flat;
  }
  return out;
}

function loadSnapshot() {
  if (!fs.existsSync(SNAPSHOT)) return null;
  try {
    return JSON.parse(fs.readFileSync(SNAPSHOT, 'utf8'));
  } catch {
    return null;
  }
}

function saveSnapshot(models, providers, pricing) {
  fs.writeFileSync(
    SNAPSHOT,
    JSON.stringify({ models, providers, pricing }, null, 2) + '\n',
    'utf8'
  );
}

function diff(oldData, newData) {
  const om = oldData.models || {};
  const nm = newData.models || {};
  const op = oldData.providers || {};
  const np = newData.providers || {};
  const opz = oldData.pricing || {};
  const npz = newData.pricing || {};

  const d = {
    added: [],
    removed: [],
    renamed: [],
    provider: [],
    active_on: [],
    active_off: [],
    categories: [],
    caps: [],
    new_providers: [],
    price_new: [],
    price_changed: [],
  };

  const oldMids = new Set(Object.keys(om));
  const newMids = new Set(Object.keys(nm));

  for (const mid of [...newMids].filter((x) => !oldMids.has(x)).sort()) {
    const m = nm[mid];
    const catList = Object.entries(m.arenas).map(([k, v]) => `${k}: ${v.join(', ')}`).join(' • ') || 'none';
    d.added.push([mid, m, catList]);
  }

  for (const mid of [...oldMids].filter((x) => !newMids.has(x)).sort()) {
    d.removed.push([mid, om[mid]]);
  }

  for (const mid of [...oldMids].filter((x) => newMids.has(x)).sort()) {
    const o = om[mid];
    const n = nm[mid];

    if (o.displayName !== n.displayName) {
      d.renamed.push([mid, o.displayName, n.displayName]);
    }
    if (o.active !== n.active) {
      (n.active ? d.active_on : d.active_off).push([mid, n.displayName]);
    }

    // Categories
    const allCats = new Set([...Object.keys(o.arenas), ...Object.keys(n.arenas)]);
    const catChanges = [];
    for (const cat of allCats) {
      const oSet = new Set(o.arenas[cat] || []);
      const nSet = new Set(n.arenas[cat] || []);
      const added = [...nSet].filter((x) => !oSet.has(x));
      const removed = [...oSet].filter((x) => !nSet.has(x));
      if (added.length > 0) catChanges.push(`➕ ${cat}: ${added.join(', ')}`);
      if (removed.length > 0) catChanges.push(`➖ ${cat}: ${removed.join(', ')}`);
    }
    if (catChanges.length > 0) {
      d.categories.push([mid, n.displayName, catChanges]);
    }

    // Capabilities
    const capChanges = [];
    for (const f of ['openSource', 'router', 'vision']) {
      if (o[f] !== n[f]) capChanges.push(`${f}: ${o[f]} ➔ ${n[f]}`);
    }
    if (o.inputModalities.join(',') !== n.inputModalities.join(',')) {
      capChanges.push(`inputModalities: ${o.inputModalities.join(', ') || 'none'} ➔ ${n.inputModalities.join(', ') || 'none'}`);
    }
    if (capChanges.length > 0) {
      d.caps.push([mid, n.displayName, capChanges]);
    }
  }

  for (const pid of Object.keys(np).filter((x) => !op[x])) {
    d.new_providers.push([pid, np[pid]]);
  }

  for (const mid of Object.keys(npz).filter((x) => !opz[x])) {
    d.price_new.push([mid, npz[mid]]);
  }
  for (const mid of Object.keys(npz).filter((x) => opz[x])) {
    const changes = [];
    const allK = new Set([...Object.keys(npz[mid]), ...Object.keys(opz[mid])]);
    for (const k of allK) {
      if (opz[mid][k] !== npz[mid][k]) {
        changes.push(`${k}: ${opz[mid][k]} ➔ ${npz[mid][k]}`);
      }
    }
    if (changes.length > 0) d.price_changed.push([mid, changes]);
  }

  return d;
}

function hasChanges(d) {
  return Object.values(d).some((arr) => arr.length > 0);
}

function buildMessage(d) {
  const now = new Date().toUTCString();
  const lines = [`🎨 DesignArena Tracker — ${now}`, ''];

  function section(title, items, render) {
    if (!items || items.length === 0) return;
    lines.push(`${title} (${items.length})`);
    for (const it of items.slice(0, 15)) {
      lines.push(...render(it));
    }
    lines.push('');
  }

  section('🆕 New models', d.added, (it) => [
    `• ${it[1].displayName} (\`${it[0]}\`)`,
    `  Provider: ${it[1].provider} | Active: ${it[1].active ? '✅' : '❌'} | Open Source: ${it[1].openSource ? '✅' : '❌'}`,
    `  Categories: ${it[2]}`,
  ]);
  section('❌ Removed models', d.removed, (it) => [
    `• ${it[1].displayName} (\`${it[0]}\`) — Provider: ${it[1].provider}`,
  ]);
  section('✏️ Name updates', d.renamed, (it) => [
    `• ${it[1]} ➔ **${it[2]}** (\`${it[0]}\`)`,
  ]);
  section('🏢 Provider updates', d.provider, (it) => [
    `• ${it[1]} (\`${it[0]}\`): ${it[2]} ➔ **${it[3]}**`,
  ]);
  section('🟢 Activated (now live)', d.active_on, (it) => [
    `• ${it[1]} (\`${it[0]}\`)`,
  ]);
  section('🔴 Deactivated (pulled)', d.active_off, (it) => [
    `• ${it[1]} (\`${it[0]}\`)`,
  ]);
  section('🏟️ Category updates', d.categories, (it) => [
    `• ${it[1]} (\`${it[0]}\`)`,
    ...it[2].map((c) => `  ${c}`),
  ]);
  section('⚡ Capability updates', d.caps, (it) => [
    `• ${it[1]} (\`${it[0]}\`)`,
    ...it[2].map((c) => `  ${c}`),
  ]);

  return lines.join('\n').trim();
}

async function notifyDiscord(msg) {
  if (!DISCORD_WEBHOOK_URL) return;
  const payload = {
    embeds: [
      {
        title: '🎨 DesignArena Tracker Update',
        description: msg.slice(0, 4000),
        color: 0x9b59b6,
        timestamp: new Date().toISOString(),
        footer: { text: 'DesignArena Tracker • designarena.ai' },
      },
    ],
  };

  try {
    await fetch(DISCORD_WEBHOOK_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
  } catch (err) {
    console.error('DesignArena discord error:', err.message);
  }
}

async function main() {
  const old = loadSnapshot();
  const data = await fetchRegistry();
  const newModels = normModels(data.models);
  const newProviders = normProviders(data.providers);
  const newPricing = normPricing(data.pricing);

  console.log(`Design models: ${Object.keys(newModels).length}, providers: ${Object.keys(newProviders).length}`);

  if (old === null) {
    saveSnapshot(newModels, newProviders, newPricing);
    console.log('Design baseline saved.');
    return;
  }

  const d = diff(old, { models: newModels, providers: newProviders, pricing: newPricing });
  if (!hasChanges(d)) {
    console.log('DesignArena: no changes.');
    return;
  }

  const msg = buildMessage(d);
  await notifyDiscord(msg);
  saveSnapshot(newModels, newProviders, newPricing);
  console.log('DesignArena snapshot updated.');
}

main().catch((err) => {
  console.error('DesignArena error:', err);
  process.exit(1);
});

